presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp

```diff
@@ -232,31 +243,31 @@ void PrestoExchangeSource::processDataResponse(
       kCounterPrestoExchangeSerializedPageSize, page ? page->size() : 0);
 
   {
-    std::vector<ContinuePromise> promises;
+    ContinuePromise requestPromise;
+    std::vector<ContinuePromise> queuePromises;
     {
       std::lock_guard<std::mutex> l(queue_->mutex());
       if (page) {
         VLOG(1) << "Enqueuing page for " << basePath_ << "/" << sequence_
                 << ": " << page->size() << " bytes";
         ++numPages_;
-        queue_->enqueueLocked(std::move(page), promises);
+        queue_->enqueueLocked(std::move(page), queuePromises);
       }
       if (complete) {
         VLOG(1) << "Enqueuing empty page for " << basePath_ << "/" << sequence_;
         atEnd_ = true;
-        queue_->enqueueLocked(nullptr, promises);
+        queue_->enqueueLocked(nullptr, queuePromises);
       }
 
       sequence_ = ackSequence;
+      requestPending_ = false;
 
-      // Reset requestPending_ if the response is complete or have pages.
-      if (complete || !empty) {
-        requestPending_ = false;
-      }
+      requestPromise = std::move(promise_);
```

Comment:
Add check here that promise_ is not empty? Thanks!

nit: requestPromise(promise_) doesn't need to be handled under the queue lock? Thanks!